AutomatticTracks/src/main/java/com/automattic/android/tracks/crashlogging/CrashLogging.kt

```diff
@@ -0,0 +1,101 @@
+
+import android.content.Context
+import com.automattic.android.tracks.BuildConfig
+import com.automattic.android.tracks.crashlogging.internal.SentryErrorTrackerProxy
+import com.automattic.android.tracks.crashlogging.internal.SentryErrorTrackerProxyImpl
+import io.sentry.SentryEvent
+import io.sentry.SentryLevel
+import io.sentry.SentryOptions
+import io.sentry.protocol.Message
+import io.sentry.protocol.User
+
+object CrashLogging {
+
+    private lateinit var dataProvider: CrashLoggingDataProvider
+    private lateinit var sentryProxy: SentryErrorTrackerProxy
+
+    @JvmStatic
+    fun start(
+        context: Context,
+        dataProvider: CrashLoggingDataProvider,
+    ) {
+        start(context, dataProvider, SentryErrorTrackerProxyImpl())
+    }
+
+    internal fun start(
+        context: Context,
+        dataProvider: CrashLoggingDataProvider,
+        sentryProxy: SentryErrorTrackerProxy
+    ) {
+        this.sentryProxy = sentryProxy
+        this.dataProvider = dataProvider
+
+        sentryProxy.initialize(context) { options ->
+            options.apply {
+                dsn = dataProvider.sentryDSN
+                environment = dataProvider.buildType
+                release = dataProvider.releaseName
+                setDebug(BuildConfig.DEBUG)
+                setTag("locale", dataProvider.locale?.language ?: "unknown")
+                beforeSend = SentryOptions.BeforeSendCallback { event, _ ->
+                    return@BeforeSendCallback if (dataProvider.userHasOptedOut) {
+                        null
+                    } else {
+                        event
+                    }
+                }
+            }
+        }
+        setNeedsDataRefresh()
+    }
+
+    @JvmStatic
+    fun setNeedsDataRefresh() {
+        applyUserTracking()
+        applyApplicationContext()
+    }
+
+    private fun applyUserTracking() {
+        sentryProxy.clearBreadcrumbs()
+
+        sentryProxy.setUser(
+            dataProvider.currentUser?.let { tracksUser ->
+                User().apply {
+                    email = tracksUser.email
+                    username = tracksUser.username
+                    others = dataProvider.userContext
+                        .plus("userID" to tracksUser.userID)
+                }
+            }
+        )
+    }
+
+    private fun applyApplicationContext() {
+        dataProvider.applicationContext.forEach { entry ->
+            sentryProxy.applyExtra(entry.key, entry.value.orEmpty())
+        }
+    }
+
+    @JvmStatic
+    fun log(throwable: Throwable) {
+        sentryProxy.captureException(throwable)
+    }
+
+    @JvmStatic
+    fun log(throwable: Throwable, data: Map<String, String?>) {
+        val event = SentryEvent().apply {
+            message = Message().apply {
+                message = throwable.message
+            }
+            level = SentryLevel.ERROR
+            setExtras(data.toMutableMap() as Map<String, String?>)
```

Comment:
This map **has** to be a mutable one, otherwise, we'll have a crash in runtime as Sentry SDK tries to edit an immutable map (which is a default in `Kotlin`).

I've reported an issue on Sentry's Github, more details and reproduction repository is there if you're interested: https://github.com/getsentry/sentry-java/issues/1357